### PR TITLE
Bug: bidirerctionalastar ignores complex restriction

### PR DESCRIPTION
**Pull request by merkispavel**
# Issue

Bug details: https://github.com/valhalla/valhalla/pull/2781#discussion_r560913655
the PR has both a gurka test which fails on master and the suggested fix

## Tasklist

 - [ ] Add tests
 - [ ] Add #fixes with the issue number that this PR addresses
 - [ ] Generally use squash merge to rebase and clean comments before merging
 - [ ] Update the [changelog](CHANGELOG.md)

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

File: test/gurka/test_route.cc [hunk 1]
```diff
@@ -416,3 +416,43 @@ TEST(Standalone, LegWeightRegression) {
...
+       }},
+  };
+
+  const auto layout = gurka::detail::map_to_coordinates(ascii_map, 100);
+  auto map = gurka::buildtiles(layout, ways, {}, relations, "test/data/dont_ignore_restriction",
+                               {{"mjolnir.concurrency", "1"}});
+
+  try {
+    auto result = gurka::route(map, "D", "E", "auto");
+    gurka::assert::raw::expect_path(result, {"Unexpected path found"});
```
Comment (merkispavel) [inline test/gurka/test_route.cc:454]: It returns `DC-CB-BE`

Comment (merkispavel) [reply test/gurka/test_route.cc:454]: From my view the bug is around this line https://github.com/valhalla/valhalla/blob/ff25b1116659d824ce5e1ecb3784b4ffe0a1e43e/src/thor/bidirectional_astar.cc#L764
`rev_pred` == `DC` and `rev_pred.on_complex_restr()` returns false

Comment (merkispavel) [reply test/gurka/test_route.cc:454]: I guess the problem is in [restrictionbuilder.cc](https://github.com/valhalla/valhalla/blob/93c955f78e1a0c61ca7898ef570d771aee9f03ce/src/mjolnir/restrictionbuilder.cc#L319). 

Let's say we have a restriction `A -> B -> C`. So we take the edge `AB` and it's marked as `start_restriction`. We go forward and reach the node `C`. We go backwards and finally have the edges `{CB, BA}`. After that we add reverse restriction using the `BA` edge as from. **BUT** we add it to the tile of the `AB` edge.
tldr: if nodes `A` and `B` belong to different tiles (Houston) we got a problem

Comment (kevinkreiser) [reply test/gurka/test_route.cc:454]: the good news is this should be incredibly rare! but also wow, good find, what a needle in a haystack. i suppose we have to detect when restrictions cross tile boundaries in the restrictions builder and patch it to handle that case.

File: src/mjolnir/restrictionbuilder.cc [hunk 2]
```diff
@@ -331,8 +329,7 @@ void build(const std::string& complex_restriction_from_file,
             if (restriction.type() < RestrictionType::kOnlyRightTurn ||
                 restriction.type() > RestrictionType::kOnlyStraightOn) {
 
-              GraphId currentNode = GraphId(tile->id().tileid(), tile->id().level(), i);
-              GraphId tileid = tile->id();
+              GraphId currentNode = directededge.endnode();
```
Comment (merkispavel) [inline src/mjolnir/restrictionbuilder.cc:332]: Turned out that both forward and it's opposite edge are marked with `start_restriction` flag.
So let's say we have a restriction `A - B - C`. Both `AB` & `BA` are `start_restriction`-s.
Before this change:
- algorithm starts with the node A and succeeds when `directededge == AB`
After this change
- algorithm starts with the node A and succeeds when `directededge == BA`

so the PR doesn't breaks anything

File: test/gurka/test_route.cc (hunk 1 above)
Comment (merkispavel) [reply test/gurka/test_route.cc:454]: > the good news is this should be incredibly rare
agreed!

**Comment by merkispavel** [comment]
Going to build large enough area to compare the number of restrictions against master to make sure the data is static

**Comment by merkispavel** [comment]
> Going to build large enough area to compare the number of restrictions against master to make sure the data is static

Did a test with `florida-latest.osm.pbf`. Suprisingly  I got 1 more restriction with my changes. Investigating 👀 ..

master
```
2021/01/25 14:07:30.819573 [INFO] Adding Restrictions at level 2
2021/01/25 14:07:42.640216 [INFO] --Forward restrictions added: 534
2021/01/25 14:07:42.640241 [INFO] --Reverse restrictions added: 345
2021/01/25 14:07:42.640766 [INFO] Adding Restrictions at level 1
2021/01/25 14:07:44.755671 [INFO] --Forward restrictions added: 1079
2021/01/25 14:07:44.755694 [INFO] --Reverse restrictions added: 1214
2021/01/25 14:07:44.755969 [INFO] Adding Restrictions at level 0
2021/01/25 14:07:46.014482 [INFO] --Forward restrictions added: 1033
2021/01/25 14:07:46.014505 [INFO] --Reverse restrictions added: 1065
```

this branch
```
2021/01/25 14:05:33.527092 [INFO] Adding Restrictions at level 2
2021/01/25 14:05:46.330749 [INFO] --Forward restrictions added: 535 // vs 534 in master
2021/01/25 14:05:46.330792 [INFO] --Reverse restrictions added: 345
2021/01/25 14:05:46.331898 [INFO] Adding Restrictions at level 1
2021/01/25 14:05:48.692780 [INFO] --Forward restrictions added: 1079
2021/01/25 14:05:48.692811 [INFO] --Reverse restrictions added: 1214
2021/01/25 14:05:48.693068 [INFO] Adding Restrictions at level 0
2021/01/25 14:05:50.047939 [INFO] --Forward restrictions added: 1033
2021/01/25 14:05:50.047961 [INFO] --Reverse restrictions added: 1065
```

**Comment by merkispavel** [comment]
> > Going to build large enough area to compare the number of restrictions against master to make sure the data is static
> 
> Did a test with `florida-latest.osm.pbf`. Suprisingly I got 1 more restriction with my changes. Investigating 👀 ..
> 
> master
> 
> ```
> 2021/01/25 14:07:30.819573 [INFO] Adding Restrictions at level 2
> 2021/01/25 14:07:42.640216 [INFO] --Forward restrictions added: 534
> 2021/01/25 14:07:42.640241 [INFO] --Reverse restrictions added: 345
> 2021/01/25 14:07:42.640766 [INFO] Adding Restrictions at level 1
> 2021/01/25 14:07:44.755671 [INFO] --Forward restrictions added: 1079
> 2021/01/25 14:07:44.755694 [INFO] --Reverse restrictions added: 1214
> 2021/01/25 14:07:44.755969 [INFO] Adding Restrictions at level 0
> 2021/01/25 14:07:46.014482 [INFO] --Forward restrictions added: 1033
> 2021/01/25 14:07:46.014505 [INFO] --Reverse restrictions added: 1065
> ```
> 
> this branch
> 
> ```
> 2021/01/25 14:05:33.527092 [INFO] Adding Restrictions at level 2
> 2021/01/25 14:05:46.330749 [INFO] --Forward restrictions added: 535 // vs 534 in master
> 2021/01/25 14:05:46.330792 [INFO] --Reverse restrictions added: 345
> 2021/01/25 14:05:46.331898 [INFO] Adding Restrictions at level 1
> 2021/01/25 14:05:48.692780 [INFO] --Forward restrictions added: 1079
> 2021/01/25 14:05:48.692811 [INFO] --Reverse restrictions added: 1214
> 2021/01/25 14:05:48.693068 [INFO] Adding Restrictions at level 0
> 2021/01/25 14:05:50.047939 [INFO] --Forward restrictions added: 1033
> 2021/01/25 14:05:50.047961 [INFO] --Reverse restrictions added: 1065
> ```

the mismatched restriction itself https://www.openstreetmap.org/relation/7680229#map=18/25.74885/-80.21806

File: src/mjolnir/restrictionbuilder.cc [hunk 3]
```diff
@@ -519,9 +513,9 @@ void build(const std::string& complex_restriction_from_file,
                     res_way_ids.push_back(restriction.to());
                   }
 
-                  tmp_ids = GetGraphIds(currentNode, reader, tileid, lock, res_way_ids);
+                  tmp_ids = GetGraphIds(currentNode, reader, lock, res_way_ids);
 
-                  if (tmp_ids.size()) {
+                  if (tmp_ids.size() && tmp_ids.back().Tile_Base() == tile_id) {
```
Comment (merkispavel) [inline src/mjolnir/restrictionbuilder.cc:518]: it happens if we split way into a few edges and not all edges are from the same tile. 
E.g. you have the following
```cpp
A -- B -- C
          | 
          D
```
There are cases that A-B-C is single osmway. And if the edges `AB` and `BC` are from the different tiles that code produced redundant restrictions: they are harmless but useless

**Comment by merkispavel** [comment]
> > Going to build large enough area to compare the number of restrictions against master to make sure the data is static
> 
> Did a test with `florida-latest.osm.pbf`. Suprisingly I got 1 more restriction with my changes. Investigating 👀 ..
> 
> UPD: the mismatched restriction https://www.openstreetmap.org/relation/7680229#map=18/25.74885/-80.21806
> 
> master
> 
> ```
> 2021/01/25 14:07:30.819573 [INFO] Adding Restrictions at level 2
> 2021/01/25 14:07:42.640216 [INFO] --Forward restrictions added: 534
> 2021/01/25 14:07:42.640241 [INFO] --Reverse restrictions added: 345
> 2021/01/25 14:07:42.640766 [INFO] Adding Restrictions at level 1
> 2021/01/25 14:07:44.755671 [INFO] --Forward restrictions added: 1079
> 2021/01/25 14:07:44.755694 [INFO] --Reverse restrictions added: 1214
> 2021/01/25 14:07:44.755969 [INFO] Adding Restrictions at level 0
> 2021/01/25 14:07:46.014482 [INFO] --Forward restrictions added: 1033
> 2021/01/25 14:07:46.014505 [INFO] --Reverse restrictions added: 1065
> ```
> 
> this branch
> 
> ```
> 2021/01/25 14:05:33.527092 [INFO] Adding Restrictions at level 2
> 2021/01/25 14:05:46.330749 [INFO] --Forward restrictions added: 535 // vs 534 in master
> 2021/01/25 14:05:46.330792 [INFO] --Reverse restrictions added: 345
> 2021/01/25 14:05:46.331898 [INFO] Adding Restrictions at level 1
> 2021/01/25 14:05:48.692780 [INFO] --Forward restrictions added: 1079
> 2021/01/25 14:05:48.692811 [INFO] --Reverse restrictions added: 1214
> 2021/01/25 14:05:48.693068 [INFO] Adding Restrictions at level 0
> 2021/01/25 14:05:50.047939 [INFO] --Forward restrictions added: 1033
> 2021/01/25 14:05:50.047961 [INFO] --Reverse restrictions added: 1065
> ```

Everything's fine. The master code used to loose a restriction. This PR now fixes it. Ready for review

**Review by PureW — APPROVED** [review]
LGTM

File: CHANGELOG.md [hunk 4]
```diff
@@ -130,6 +130,7 @@
    * FIXED: Remove redundant return statements [#2776](https://github.com/valhalla/valhalla/pull/2776)
    * FIXED: Add support for geos-3.9 c++ api [#2739](https://github.com/valhalla/valhalla/issues/2739)
    * FIXED: Fix check for live speed validness [#2797](https://github.com/valhalla/valhalla/pull/2797)
+   * FIXED: Store restrictions in the right tile [#2781](https://github.com/valhalla/valhalla/pull/2781)
```
Comment (kevinkreiser) [inline CHANGELOG.md:133]: we made a release please move this to the correct section above

**Review by PureW — APPROVED** [review]
